### PR TITLE
LancedDB integration

### PR DESCRIPTION
**Pull request by gsilvestrin**
- LanceDBVectorStore provides data manipulation and query capabilities on top of LanceDB.
- GPTLanceDBIndex exposes an index backed by a LanceDBVectorStore

### CODE REVIEW

File: gpt_index/indices/vector_store/vector_indices.py [hunk 1]
```diff
@@ -350,6 +352,42 @@ def __init__(
...
+class GPTLanceDBIndex(GPTVectorStoreIndex):
+    index_struct_cls: Type[IndexDict] = LanceDBIndexDict
+
+    def __init__(
+        self,
+        nodes: Optional[Sequence[Node]] = None,
+        uri: Optional[str] = None,
+        table_name: str = "vectors",
+        id_column_name: str = "id",
+        nprobes: int = 10,
```
Comment (changhiskhan) [inline gpt_index/indices/vector_store/vector_indices.py]: How come this default is 10 but it's 20 elsewhere

File: gpt_index/indices/vector_store/vector_indices.py [hunk 2]
```diff
@@ -350,6 +352,42 @@ def __init__(
             **kwargs,
         )
 
+class GPTLanceDBIndex(GPTVectorStoreIndex):
+    index_struct_cls: Type[IndexDict] = LanceDBIndexDict
+
+    def __init__(
```
Comment (changhiskhan) [inline gpt_index/indices/vector_store/vector_indices.py:382]: Need comments to describe the parameters?

File: gpt_index/vector_stores/lancedb.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    VectorStoreQuery,
+)
+
+
+class LanceDBVectorStore(VectorStore):
+    """The LanceDB Vector Store
+
+    The embeddings are stored in LanceDB.
+
+    During query time,
```
Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py]: ?

File: gpt_index/indices/vector_store/vector_indices.py [hunk 4]
```diff
@@ -350,6 +352,42 @@ def __init__(
             **kwargs,
         )
 
+class GPTLanceDBIndex(GPTVectorStoreIndex):
```
Comment (changhiskhan) [inline gpt_index/indices/vector_store/vector_indices.py:356]: What does the index do?

File: gpt_index/vector_stores/lancedb.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            tbl.add(data)
+        else:
+            self.connection.create_table(self.table_name, data)
+        return ids
+
+    def query(
+            self,
+            query: VectorStoreQuery,
+    ) -> VectorStoreQueryResult:
+        table = self.connection.open_table(self.table_name)
```
Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py:116]: Should we cache this so Lance metadata doesn't have to be read on every query? Also if we add caching we need a stable Table instance prolly

Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py:116]: Should we cache this so Lance metadata doesn't have to be read on every query? Also if we add caching we need a stable Table instance prolly

File: gpt_index/vector_stores/lancedb.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        query = table.search(query.query_embedding) \
+            .limit(query.similarity_top_k) \
+            .nprobes(self.nprobes)
+
+        if self.refine_factor is not None:
+            query.refine_factor(self.refine_factor)
+
+        results = query.to_df()
```
Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py]: I guess we should also add a to_arrow method so we don't need to pay the extra conversion cost to pandas. Not important for now

File: gpt_index/vector_stores/lancedb.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            ids.append(result.id)
+
+        if "vectors" in self.connection.table_names():
+            tbl = self.connection.open_table("vectors")
+            tbl.add(data)
+        else:
+            self.connection.create_table(self.table_name, data)
+        return ids
+
+    def query(
```
Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py:111]: nprobes and refine_factor should be query time parameters - is there a way for the user to override the defaults?

File: gpt_index/vector_stores/lancedb.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        for result in embedding_results:
+            data.append({
+                "id": result.id,
+                "doc_id": result.doc_id,
+                "vector": result.embedding,
+                "text": result.node.get_text(),
+            })
+            ids.append(result.id)
+
+        if "vectors" in self.connection.table_names():
```
Comment (changhiskhan) [inline gpt_index/vector_stores/lancedb.py]: shouldn't this be self.table_name?

**Review by changhiskhan — COMMENTED** [review]
Looking at this code I don't think it makes sense to create the index automatically since embeddings are added incrementally. But at some point the user needs to create an index. I don't know what the right UX is for that.

File: gpt_index/vector_stores/lancedb.py (hunk 5 above)
Comment (gsilvestrin) [reply gpt_index/vector_stores/lancedb.py:116]: do you mean just use `self.table = ...` or something more complicated?

File: gpt_index/vector_stores/lancedb.py (hunk 7 above)
Comment (gsilvestrin) [reply gpt_index/vector_stores/lancedb.py:111]: Unfortunately not at query time - but you can always create multiple indexes that point to to the same lance data file

File: gpt_index/indices/vector_store/vector_indices.py (hunk 2 above)
Comment (gsilvestrin) [reply gpt_index/indices/vector_store/vector_indices.py:382]: Added

File: gpt_index/indices/vector_store/vector_indices.py (hunk 4 above)
Comment (gsilvestrin) [reply gpt_index/indices/vector_store/vector_indices.py:356]: Added comments

File: gpt_index/indices/vector_store/vector_indices.py (hunk 1 above)
Comment (gsilvestrin) [reply gpt_index/indices/vector_store/vector_indices.py]: It was wrong, I fixed it

File: gpt_index/vector_stores/lancedb.py (hunk 3 above)
Comment (gsilvestrin) [reply gpt_index/vector_stores/lancedb.py]: Added comments

File: gpt_index/vector_stores/lancedb.py (hunk 8 above)
Comment (gsilvestrin) [reply gpt_index/vector_stores/lancedb.py]: it should - fixed it

**Pull request closed without merging**